CNF-15733:kubeletcontroller:hypershift: configmap cleanup

Tal-or opened the pull request:
On hypershift we cannot rely on the automatic garbage collection mechanism provided by k8s controllers.

The reason is becuase we can not set the owner reference across namespaces.

Hence, we need to add some custom logic to cleanup the ConfigMap generated for RTE.

Tal-or committed:
kubeletcontroller:hypershift: configmap cleanup

On hypershift we cannot rely on the automatic garbage collection
mechanism provided by k8s controllers.

The reason is becuase we can not set the owner reference
across namespaces.

Hence, we need to add some custom logic to cleanup the
ConfigMap generated for RTE.

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

ffromani commented on controllers/kubeletconfig_controller.go:
```diff
@@ -140,6 +144,15 @@ func (r *KubeletConfigReconciler) SetupWithManager(mgr ctrl.Manager, clusterPlat
+				return false
+			}
+			klog.InfoS("KubeletConfig ConfigMap object got deleted", "KubeletConfig", kubelet.Name)
+			r.GarbageCollectionFor = append(r.GarbageCollectionFor, *kubelet)
```
maybe a deepcopy?

Tal-or replied:
`DeepCopy` still return a pointer I can make this a slice of pointers and save some redundant copying

ffromani commented on controllers/kubeletconfig_controller.go:
```diff
@@ -66,6 +67,9 @@ type KubeletConfigReconciler struct {
 	Platform  platform.Platform
+	// GarbageCollectionFor is a list of objects
+	// that their dependent needs to be removed from the cluster.
+	GarbageCollectionFor []corev1.ConfigMap
```
does this need to be public member?

Tal-or replied:
No. Let's change it

ffromani reviewed (commented):
looks better than I anticipated. How hard is to actually run the logic only if running on hypershift?

Tal-or commented:
> looks better than I anticipated. How hard is to actually run the logic only if running on hypershift?

The interception of the deletion event is done only for HyperShift predicates

ffromani commented:
> > looks better than I anticipated. How hard is to actually run the logic only if running on hypershift?
> 
> The interception of the deletion event is done only for HyperShift predicates

Right, I should have been more careful in reading the code :P

ffromani commented on controllers/kubeletconfig_controller.go:
```diff
@@ -275,6 +295,51 @@ func (r *KubeletConfigReconciler) makeKCHandlerForPlatform(ctx context.Context,
+func getDeletedOwner(kcKey client.ObjectKey, ownerConfigMaps []corev1.ConfigMap) *corev1.ConfigMap {
+	for i := range ownerConfigMaps {
+		cm := &ownerConfigMaps[i]
+		if client.ObjectKeyFromObject(cm).String() == kcKey.String() {
```
let's check if there's a recommended way to compare keys. I don't see obvious issues with this one, but still

Tal-or committed:
e2e: check configmap cleanup

Check that if we're deleting the kubeletconfig
the derived configmap that nrop generated from it
has been removed as well.

This kind of check unique for hypershift,
since we cannot set the configmap owner reference
to be the kubeletconfig configmap because it lies
on a different namespace.

Therefore, we need to validate that our custom cleanup
logic is working.

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

Tal-or commented on test/utils/deploy/hypershift.go:
```diff
@@ -91,6 +95,25 @@ func (h *HyperShiftNRO) Teardown(ctx context.Context, timeout time.Duration) {
 
 		Expect(e2eclient.MNGClient.Delete(ctx, h.KcConfigMapObj)).To(Succeed())
+
+		By("checking that generated configmap has been deleted")
```
IMO this is not the ideal place for checking the deletion. 
It would be more appropriate in the `uninstall` suite.
But after checking the `uninstall` suite I see it completely redundant.
We're deleting the nro and kubeletconfig in the `uninstall` suite but these objects were already removed here:
https://github.com/openshift-kni/numaresources-operator/blob/7de8e4b52dcff68106fa0efa8f5b218acfdf423a/test/e2e/install/install_test.go#L203 

In the nutshell the `uninstall` need to go through some refactoring but this is another issue for another PR.

ffromani replied:
make sense to me with the catch that suites should not depend on each other as much as possible; IOW they should be indpedent and self-sufficient.

Tal-or commented:
/hold cancel
This PR is ready

ffromani commented on controllers/kubeletconfig_controller.go:
```diff
@@ -275,6 +295,51 @@ func (r *KubeletConfigReconciler) makeKCHandlerForPlatform(ctx context.Context,
+	return nil
+}
+
+func removeDeletedOwner(kcKey client.ObjectKey, ownerConfigMaps []*corev1.ConfigMap) []*corev1.ConfigMap {
+	for i := range ownerConfigMaps {
+		cm := ownerConfigMaps[i]
+		if client.ObjectKeyFromObject(cm) == kcKey {
+			return append(ownerConfigMaps[:i], ownerConfigMaps[i+1:]...)
+		}
+	}
+	return ownerConfigMaps
+}
```
This does not look obviously wrong but makes me wary. I'd go for the trivialest implementation:
```
func removeDeletedOwner(kcKey client.ObjectKey, ownerConfigMaps []*corev1.ConfigMap) []*corev1.ConfigMap {
        ret := make([]*corev1.ConfigMap, 0, len(ownerConfigMaps))
	for i := range ownerConfigMaps {
		cm := ownerConfigMaps[i]
		if client.ObjectKeyFromObject(cm) == kcKey {
			continue
		}
                ret = append(ret, cm)
	}
	return ret
}
```
perhaps something in the `slices` package can help?

ffromani replied:
maybe https://pkg.go.dev/slices#DeleteFunc ?

Tal-or replied:
Nice. I used https://pkg.go.dev/slices#Delete which is good enough

ffromani commented on controllers/kubeletconfig_controller.go:
```diff
@@ -275,6 +296,50 @@ func (r *KubeletConfigReconciler) makeKCHandlerForPlatform(ctx context.Context,
+func removeDeletedOwner(kcKey client.ObjectKey, ownerConfigMaps []*corev1.ConfigMap) []*corev1.ConfigMap {
+	for i := range ownerConfigMaps {
+		if client.ObjectKeyFromObject(ownerConfigMaps[i]) == kcKey {
+			return slices.Delete(ownerConfigMaps, i, i+1)
```
ok, let's start like this for now

ffromani reviewed (commented):
/approve
/lgtm
